In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import matplotlib.pyplot as plt
import warnings
import seaborn as sns

url = "https://raw.githubusercontent.com/paulopadilla/digital_mkt/main/taller3/boletas_20000.csv"
data = pd.read_csv(url, sep=';')


url_margen = "https://raw.githubusercontent.com/paulopadilla/digital_mkt/main/taller3/margen_categorias.csv"



# eliminar datos duplicados
num_duplicadas = data.duplicated().sum()
data.drop_duplicates(inplace=True)

# Verificar y contar valores nulos
nulos = data.isnull().sum().sum()

# Imprimir resultados
print(f"\n Filas duplicadas eliminadas: {num_duplicadas if num_duplicadas > 0 else '0'}")
if data.isnull().any().any():
  print("\n Hay datos nulos en la BBDD, aquí el desglose: \n", data.isnull().sum())
else:
  print ("\n No hay datos nulos en la BBDD\n")

print("Info de los datos:")
print(data.info())
print(data.head())

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)



 Filas duplicadas eliminadas: 0

 No hay datos nulos en la BBDD

Info de los datos:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19999 entries, 0 to 19998
Columns: 279 entries, Etiquetas de fila to YOGHURT
dtypes: int64(279)
memory usage: 42.6 MB
None
   Etiquetas de fila   CUADERNO ESCOLAR  ACEITE DE MARAVILLA  ACEITE DE OLIVA  \
0           60888573                  0                    0                0   
1           60888574                  0                    0                0   
2           60888575                  0                    0                0   
3           60888576                  0                    0                0   
4           60888577                  0                    0                0   

   ACEITE PEPITA DE UVA  ACEITE VEGETAL  ACEITUNAS  ADHESIVOS  AGUARDIENTE  \
0                     0               0          0          0            0   
1                     0               1          0          0            0   
2                    

In [ ]:
print("Dimensión de la tabla: ", data.shape)

Dimensión de la tabla:  (19999, 279)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


# Pregunta 1
#### Corra el algoritimo apriori sobre la BD transaccional usando 1000, 100 y 10 transacciones como soporte mínimo. Reporte el tiempo de procesamiento y número de reglas extraídas. Explique el resultado.

In [ ]:
#1000 transacciones mínimas
soporte_min_1000 = 1000/data.shape[0]
frequent_itemsets_1000 = apriori(data.iloc[:, 2:], min_support = soporte_min_1000, use_colnames = True, max_len = 2)
display(frequent_itemsets_1000)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                     itemsets
0    0.162408             (ACEITE VEGETAL)
1    0.153458              (ARROZ REGULAR)
2    0.084004                       (ATUN)
3    0.180459           (AZUCAR GRANULADA)
4    0.072254              (BEBIDA LACTEA)
..        ...                          ...
144  0.098405  (SALSAS DE TOMATE, YOGHURT)
145  0.058103   (YOGHURT, SIN DESCRIPCION)
146  0.066303       (TE EN BOLSA, YOGHURT)
147  0.055753  (YOGHURT, TOALLAS DE PAPEL)
148  0.068753          (VIENESAS, YOGHURT)

[149 rows x 2 columns]

In [ ]:
rules_1000 = association_rules(frequent_itemsets_1000, metric="lift", min_threshold = 1)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
#100 transacciones mínimas
soporte_min_100 = 100/data.shape[0]
frequent_itemsets_100 = apriori(data.iloc[:, 2:], min_support = soporte_min_100, use_colnames = True, max_len = 2)
display(frequent_itemsets_100)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                      itemsets
0     0.025301         (ACEITE DE MARAVILLA)
1     0.012001             (ACEITE DE OLIVA)
2     0.162408              (ACEITE VEGETAL)
3     0.047102                       (AGUAS)
4     0.009550                         (AJI)
...        ...                           ...
3075  0.008050             (VINOS, VIENESAS)
3076  0.068753           (VIENESAS, YOGHURT)
3077  0.008050           (VINAGRES, YOGHURT)
3078  0.015001              (VINOS, YOGHURT)
3079  0.007150  (YERBA MATE GRANEL, YOGHURT)

[3080 rows x 2 columns]

In [ ]:
rules_100 = association_rules(frequent_itemsets_100, metric="lift", min_threshold = 1)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
#10 transacciones mínimas
soporte_min_10 = 10/data.shape[0]
frequent_itemsets_10 = apriori(data.iloc[:, 2:], min_support = soporte_min_10, use_colnames = True, max_len = 2)
display(frequent_itemsets_10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                      itemsets
0      0.025301         (ACEITE DE MARAVILLA)
1      0.012001             (ACEITE DE OLIVA)
2      0.002750        (ACEITE PEPITA DE UVA)
3      0.162408              (ACEITE VEGETAL)
4      0.002750                   (ACEITUNAS)
...         ...                           ...
12654  0.001350    (VINOS, YERBA MATE GRANEL)
12655  0.015001              (VINOS, YOGHURT)
12656  0.001650         (VIRUTILLAS, YOGHURT)
12657  0.000500             (WHISKY, YOGHURT)
12658  0.007150  (YERBA MATE GRANEL, YOGHURT)

[12659 rows x 2 columns]

In [ ]:
rules_10 = association_rules(frequent_itemsets_10, metric="lift", min_threshold = 1)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Se puede observar que con la condición de 1000 transacciones mínimas se encontraron 149 reglas y el tiempo de ejecución fue de 1 segundo. Para 100 transacciones minimas se obtuvieron 3080 reglas y el tiempo de ejecución fue de 2 segundos. Por último, para 10 transacciones se encontraron 12659 reglas y se demoró 21 segundos. Este cambio se debe a que, mientras la condición baje, se van a encontrar más transacciones que cumplan, por lo que la tarea se demorará más en terminar.

# Pregunta 2
#### Vuelva a correr a priori usando 20 transacciones como soporte mínimo y reporte las 20 reglas entre dos categorías, con mayor soporte, confianza y ganancia. Interprete y explique los valores para la regla con mayor soporte, confianza y ganancia.

In [ ]:
soporte_min_20 = 20/data.shape[0]
frequent_itemsets_20 = apriori(data.iloc[:, 2:], min_support = soporte_min_20, use_colnames = True, max_len = 2)
display(frequent_itemsets_20)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


support                      itemsets
0     0.025301         (ACEITE DE MARAVILLA)
1     0.012001             (ACEITE DE OLIVA)
2     0.002750        (ACEITE PEPITA DE UVA)
3     0.162408              (ACEITE VEGETAL)
4     0.002750                   (ACEITUNAS)
...        ...                           ...
9113  0.008050           (VINAGRES, YOGHURT)
9114  0.001350    (VINOS, YERBA MATE GRANEL)
9115  0.015001              (VINOS, YOGHURT)
9116  0.001650         (VIRUTILLAS, YOGHURT)
9117  0.007150  (YERBA MATE GRANEL, YOGHURT)

[9118 rows x 2 columns]

In [ ]:
rules_20 = association_rules(frequent_itemsets_20, metric="lift", min_threshold = 1)
print("Número de reglas: ", rules_20.shape[0])

rules_20.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Número de reglas:  17640


antecedents            consequents  antecedent support  \
0  (ACEITE DE MARAVILLA)      (ACEITE DE OLIVA)            0.025301   
1      (ACEITE DE OLIVA)  (ACEITE DE MARAVILLA)            0.012001   
2  (ACEITE DE MARAVILLA)       (ACEITE VEGETAL)            0.025301   
3       (ACEITE VEGETAL)  (ACEITE DE MARAVILLA)            0.162408   
4  (ACEITE DE MARAVILLA)                (AGUAS)            0.025301   

   consequent support  support  confidence      lift  leverage  conviction  \
0            0.012001  0.00140    0.055336  4.611100  0.001096    1.045874   
1            0.025301  0.00140    0.116667  4.611100  0.001096    1.103433   
2            0.162408  0.00460    0.181818  1.119514  0.000491    1.023723   
3            0.025301  0.00460    0.028325  1.119514  0.000491    1.003112   
4            0.047102  0.00235    0.092885  1.971990  0.001158    1.050471   

   zhangs_metric  
0       0.803461  
1       0.792644  
2       0.109526  
3       0.127455  
4       0.505693

In [ ]:
rules_20_soporte = rules_20.sort_values('support', ascending= False)
rules_20_soporte.head(20)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
11132    (GALLETAS DULCES)            (YOGHURT)            0.255163   
11133            (YOGHURT)    (GALLETAS DULCES)            0.306515   
14057    (LECHES LIQUIDAS)            (YOGHURT)            0.231312   
14056            (YOGHURT)    (LECHES LIQUIDAS)            0.306515   
10346        (FIDEO CORTO)   (SALSAS DE TOMATE)            0.159108   
10347   (SALSAS DE TOMATE)        (FIDEO CORTO)            0.171559   
10576        (FIDEO LARGO)   (SALSAS DE TOMATE)            0.154708   
10577   (SALSAS DE TOMATE)        (FIDEO LARGO)            0.171559   
10956    (GALLETAS DULCES)    (LECHES LIQUIDAS)            0.255163   
10957    (LECHES LIQUIDAS)    (GALLETAS DULCES)            0.231312   
10165        (FIDEO LARGO)        (FIDEO CORTO)            0.154708   
10164        (FIDEO CORTO)        (FIDEO LARGO)            0.159108   
15684  (NECTARES LIQUIDOS)            (YOGHURT)            0.178859   
15685            (YOGHURT)  (NECTARES LIQUIDOS)            0.306515   
17204   (SALSAS DE TOMATE)            (YOGHURT)            0.171559   
17205            (YOGHURT)   (SALSAS DE TOMATE)            0.306515   
413       (ACEITE VEGETAL)   (AZUCAR GRANULADA)            0.162408   
412     (AZUCAR GRANULADA)     (ACEITE VEGETAL)            0.180459   
3237    (AZUCAR GRANULADA)   (SALSAS DE TOMATE)            0.180459   
3236    (SALSAS DE TOMATE)   (AZUCAR GRANULADA)            0.171559   

       consequent support   support  confidence      lift  leverage  \
11132            0.306515  0.133507    0.523222  1.707000  0.055295   
11133            0.255163  0.133507    0.435563  1.707000  0.055295   
14057            0.306515  0.132607    0.573281  1.870319  0.061706   
14056            0.231312  0.132607    0.432626  1.870319  0.061706   
10346            0.171559  0.107055    0.672847  3.921968  0.079759   
10347            0.159108  0.107055    0.624016  3.921968  0.079759   
10576            0.171559  0.106505    0.688429  4.012794  0.079964   
10577            0.154708  0.106505    0.620810  4.012794  0.079964   
10956            0.231312  0.105455    0.413286  1.786708  0.046433   
10957            0.255163  0.105455    0.455901  1.786708  0.046433   
10165            0.159108  0.104055    0.672592  4.227269  0.079440   
10164            0.154708  0.104055    0.653991  4.227269  0.079440   
15684            0.306515  0.098805    0.552418  1.802253  0.043982   
15685            0.178859  0.098805    0.322349  1.802253  0.043982   
17204            0.306515  0.098405    0.573594  1.871338  0.045820   
17205            0.171559  0.098405    0.321044  1.871338  0.045820   
413              0.180459  0.096155    0.592057  3.280837  0.066847   
412              0.162408  0.096155    0.532835  3.280837  0.066847   
3237             0.171559  0.095805    0.530895  3.094541  0.064845   
3236             0.180459  0.095805    0.558438  3.094541  0.064845   

       conviction  zhangs_metric  
11132    1.454522       0.556064  
11133    1.319610       0.597240  
14057    1.625157       0.605358  
14056    1.354819       0.671005  
10346    2.532277       0.885995  
10347    2.236512       0.899310  
10576    2.658919       0.888210  
10577    2.229207       0.906277  
10956    1.310159       0.591151  
10957    1.368938       0.572809  
10165    2.568332       0.903168  
10164    2.442979       0.907894  
15684    1.549403       0.542098  
15685    1.211746       0.641887  
17204    1.626347       0.562047  
17205    1.220170       0.671425  
413      2.008958       0.829998  
412      1.792924       0.848279  
3237     1.766004       0.825889  
3236     1.856003       0.817016

Al ordenar según el mayor soporte, el primer lugar corresponde a yoghurt y galletas dulces. Esto significa que, dentro de las compras analizadas, un 0.133507 (o un 13%) contiene yoghurt y galletas dulces.

In [ ]:
rules_20_confianza = rules_20.sort_values('confidence', ascending= False)
rules_20_confianza.head(20)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
8760                 (COMINO)  (SALSAS DE TOMATE)            0.001450   
3824    (BEBIDA LACTEA POLVO)  (SALSAS DE TOMATE)            0.001700   
13405               (LASAGNA)  (SALSAS DE TOMATE)            0.012701   
16828  (POSTRES REFRIGERADOS)           (YOGHURT)            0.125006   
12690                 (JAMON)            (QUESOS)            0.004750   
6988               (CEREALES)           (YOGHURT)            0.070704   
6450        (CERAL AZUCARADO)           (YOGHURT)            0.016051   
14117              (LENTEJAS)  (SALSAS DE TOMATE)            0.022451   
16612                  (PAVO)           (YOGHURT)            0.003900   
3821          (BEBIDA LACTEA)           (YOGHURT)            0.072254   
16286           (PAPEL ALUSA)           (YOGHURT)            0.012601   
17460   (SURTIDO DE MARISCOS)           (YOGHURT)            0.005950   
10576           (FIDEO LARGO)  (SALSAS DE TOMATE)            0.154708   
11586      (Grupo Art. DIPAC)           (YOGHURT)            0.010251   
16741               (POROTOS)  (SALSAS DE TOMATE)            0.013251   
11398             (GARBANZOS)           (YOGHURT)            0.006500   
10152      (FIAMBRE DE JAMON)            (QUESOS)            0.004900   
10555               (POROTOS)       (FIDEO LARGO)            0.013251   
3822    (BEBIDA LACTEA POLVO)       (FIDEO CORTO)            0.001700   
3826    (BEBIDA LACTEA POLVO)           (YOGHURT)            0.001700   

       consequent support   support  confidence      lift  leverage  \
8760             0.171559  0.001150    0.793103  4.622931  0.000901   
3824             0.171559  0.001300    0.764706  4.457404  0.001008   
13405            0.171559  0.009650    0.759843  4.429056  0.007472   
16828            0.306515  0.093305    0.746400  2.435115  0.054988   
12690            0.151408  0.003500    0.736842  4.866613  0.002781   
6988             0.306515  0.051803    0.732673  2.390332  0.030131   
6450             0.306515  0.011201    0.697819  2.276621  0.006281   
14117            0.171559  0.015601    0.694878  4.050380  0.011749   
16612            0.306515  0.002700    0.692308  2.258640  0.001505   
3821             0.306515  0.049952    0.691349  2.255514  0.027806   
16286            0.306515  0.008700    0.690476  2.252664  0.004838   
17460            0.306515  0.004100    0.689076  2.248095  0.002276   
10576            0.171559  0.106505    0.688429  4.012794  0.079964   
11586            0.306515  0.007050    0.687805  2.243949  0.003908   
16741            0.171559  0.009100    0.686792  4.003253  0.006827   
11398            0.306515  0.004450    0.684615  2.233544  0.002458   
10152            0.151408  0.003350    0.683673  4.515451  0.002608   
10555            0.154708  0.009000    0.679245  4.390506  0.006950   
3822             0.159108  0.001150    0.676471  4.251645  0.000880   
3826             0.306515  0.001150    0.676471  2.206971  0.000629   

       conviction  zhangs_metric  
8760     4.004134       0.784825  
3824     3.520876       0.776975  
13405    3.449576       0.784178  
16828    2.734561       0.673538  
12690    3.224651       0.798310  
6988     2.594146       0.625901  
6450     2.294934       0.569900  
14117    2.715111       0.770406  
16612    2.253825       0.559438  
3821     2.246828       0.599994  
16286    2.240489       0.563178  
17460    2.230397       0.558502  
10576    2.658919       0.888210  
11586    2.221318       0.560098  
16741    2.645024       0.760277  
11398    2.198854       0.555894  
10152    2.682647       0.782372  
10555    2.635323       0.782606  
3822     2.599121       0.766099  
3826     2.143498       0.547822

Para la regla con mayor confianza se encuentra comino y salsa de tomate con un valor de 0.793103. Por lo tanto, dentro de las compras que llevan comino, un 79% también lleva salsa de tomate. Es importante destacar que la salsa de tomate aparece 6 veces como consecuente dentro de las 20 reglas, por lo que este resultado indica que es un producto común. Probablemente no hay una gran cantidad de compras que incluyan comino, resultando en que muchas de estas compras coincidentalmente también incluyen salsa de tomate.

In [ ]:
rules_20_ganancia = rules_20.sort_values('lift', ascending= False)
rules_20_ganancia.head(20)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents             consequents  antecedent support  \
12321    (IMPLEM  CUMPLEANOS)   (VARIOS UTENS COCINA)            0.008450   
12320   (VARIOS UTENS COCINA)    (IMPLEM  CUMPLEANOS)            0.007150   
3321                    (RON)             (BAJATIVOS)            0.011151   
3320              (BAJATIVOS)                   (RON)            0.004300   
3468                (BALSAMO)               (SHAMPOO)            0.016451   
3469                (SHAMPOO)               (BALSAMO)            0.028951   
11370             (GARBANZOS)               (POROTOS)            0.006500   
11371               (POROTOS)             (GARBANZOS)            0.013251   
7229            (CHAMPINONES)               (PALMITO)            0.005000   
7228                (PALMITO)           (CHAMPINONES)            0.015951   
12736    (SALCHICHON CERVEZA)              (JAMONADA)            0.012851   
12737              (JAMONADA)    (SALCHICHON CERVEZA)            0.005750   
14102              (LENTEJAS)               (POROTOS)            0.022451   
14103               (POROTOS)              (LENTEJAS)            0.013251   
1067                    (AJI)               (MOSTAZA)            0.009550   
1066                (MOSTAZA)                   (AJI)            0.020901   
7310                (CHUPETE)  (CHOCLATES CONFITADOS)            0.017051   
7311   (CHOCLATES CONFITADOS)               (CHUPETE)            0.011251   
14220              (MANTECAS)   (LEVADURA PANIFICADA)            0.014301   
14221   (LEVADURA PANIFICADA)              (MANTECAS)            0.012501   

       consequent support  support  confidence       lift  leverage  \
12321            0.007150  0.00165    0.195266  27.308603  0.001590   
12320            0.008450  0.00165    0.230769  27.308603  0.001590   
3321             0.004300  0.00100    0.089686  20.856189  0.000952   
3320             0.011151  0.00100    0.232558  20.856189  0.000952   
3468             0.028951  0.00980    0.595745  20.577371  0.009324   
3469             0.016451  0.00980    0.338515  20.577371  0.009324   
11370            0.013251  0.00175    0.269231  20.318287  0.001664   
11371            0.006500  0.00175    0.132075  20.318287  0.001664   
7229             0.015951  0.00155    0.310000  19.434765  0.001470   
7228             0.005000  0.00155    0.097179  19.434765  0.001470   
12736            0.005750  0.00120    0.093385  16.240095  0.001126   
12737            0.012851  0.00120    0.208696  16.240095  0.001126   
14102            0.013251  0.00480    0.213808  16.135681  0.004503   
14103            0.022451  0.00480    0.362264  16.135681  0.004503   
1067             0.020901  0.00315    0.329843  15.781169  0.002951   
1066             0.009550  0.00315    0.150718  15.781169  0.002951   
7310             0.011251  0.00295    0.173021  15.378833  0.002758   
7311             0.017051  0.00295    0.262222  15.378833  0.002758   
14220            0.012501  0.00265    0.185315  14.824434  0.002471   
14221            0.014301  0.00265    0.212000  14.824434  0.002471   

       conviction  zhangs_metric  
12321    1.233762       0.971592  
12320    1.289014       0.970320  
3321     1.093798       0.962788  
3320     1.288501       0.956164  
3468     2.402067       0.967316  
3469     1.486880       0.979769  
11370    1.350289       0.957004  
11371    1.144684       0.963551  
7229     1.426158       0.953313  
7228     1.102100       0.963921  
12736    1.096662       0.950640  
12737    1.247496       0.943851  
14102    1.255100       0.959569  
14103    1.532843       0.950622  
1067     1.460999       0.945665  
1066     1.166219       0.956628  
7310     1.195615       0.951194  
7311     1.332311       0.945614  
14220    1.212124       0.946073  
14221    1.250887       0.944349

En el caso de la mayor ganancia, implementos de cumpleaños y varios utencilios de cocina obtienen el mayor valor, con 27.308603. Esto significa que cuando alguien lleva implementos de cumpleaños, es 27.3 veces más probable que también lleve utencilios de cocina, y viceversa.

# Pregunta 3
#### Defina tres categorías a recomendar, cuando los clientes estén visitando las páginas de MAYONESA, GALLETAS DULCES, VERDURAS CONGELADAS, SALSAS DE TOMATE. Explique qué criterios usaría para priorizar qué regla o categoría recomendar.

In [ ]:
margenes = pd.read_csv("https://raw.githubusercontent.com/paulopadilla/digital_mkt/main/taller3/margen_categorias.csv")
margenes= margenes.sort_values(by='Margen promedio', ascending=False)
print(margenes)



# eliminar datos duplicados
num_duplicadas = margenes.duplicated().sum()
margenes.drop_duplicates(inplace=True)

# Verificar y contar valores nulos
nulos = margenes.isnull().sum().sum()

# Imprimir resultados
print(f"\n Filas duplicadas eliminadas: {num_duplicadas if num_duplicadas > 0 else '0'}")
if margenes.isnull().any().any():
  print("\n Hay datos nulos en la BBDD, aquí el desglose: \n", margenes.isnull().sum())
else:
  print ("\n No hay datos nulos en la BBDD\n")

print("Info de los datos:")
print(margenes.info())
print(margenes.head())

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


                Categoria  Margen promedio
139    JUGOS CONCENTRADOS             1998
185                 OTROS             1990
94     DESODORANTE EN GEL             1983
142                 JUREL             1977
188  Otros NO Clasificado             1964
..                    ...              ...
175            MERMELADAS              127
133        JABON EN BARRA              124
209          PASTA DENTAL              122
73                  CHUNO              121
120                 GOMAS              108

[278 rows x 2 columns]

 Filas duplicadas eliminadas: 0

 No hay datos nulos en la BBDD

Info de los datos:
<class 'pandas.core.frame.DataFrame'>
Index: 278 entries, 139 to 120
Data columns (total 2 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Categoria        278 non-null    object
 1   Margen promedio  278 non-null    int64 
dtypes: int64(1), object(1)
memory usage: 6.5+ KB
None
                Categoria  Margen prome

In [ ]:
rules_100['antecedents'].dtypes
rules_100['antecedents'] = rules_100['antecedents'].astype(str)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


## Mayonesa

In [ ]:
#se filtra la tabla para que sólo muestre los antecedentes necesarios
mayonesa = rules_100[rules_100['antecedents'].str.contains('MAYONESA')]
mayonesa

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents            consequents  antecedent support  \
41    frozenset({'MAYONESA'})  (ACEITE DE MARAVILLA)            0.083804   
223   frozenset({'MAYONESA'})       (ACEITE VEGETAL)            0.083804   
349   frozenset({'MAYONESA'})                (AGUAS)            0.083804   
461   frozenset({'MAYONESA'})    (ARROZ PREGRANEADO)            0.083804   
621   frozenset({'MAYONESA'})        (ARROZ REGULAR)            0.083804   
...                       ...                    ...                 ...   
5045  frozenset({'MAYONESA'})   (TOALLAS HIGIENICAS)            0.083804   
5047  frozenset({'MAYONESA'})  (VERDURAS CONGELADAS)            0.083804   
5049  frozenset({'MAYONESA'})             (VIENESAS)            0.083804   
5051  frozenset({'MAYONESA'})                (VINOS)            0.083804   
5053  frozenset({'MAYONESA'})              (YOGHURT)            0.083804   

      consequent support   support  confidence      lift  leverage  \
41              0.025301  0.005550    0.066229  2.617621  0.003430   
223             0.162408  0.041552    0.495823  3.052947  0.027942   
349             0.047102  0.006200    0.073986  1.570743  0.002253   
461             0.034602  0.010051    0.119928  3.465965  0.007151   
621             0.153458  0.036452    0.434964  2.834425  0.023591   
...                  ...       ...         ...       ...       ...   
5045            0.037752  0.010001    0.119332  3.160948  0.006837   
5047            0.083704  0.019901    0.237470  2.837017  0.012886   
5049            0.118356  0.033552    0.400358  3.382661  0.023633   
5051            0.041952  0.007100    0.084726  2.019578  0.003585   
5053            0.306515  0.045852    0.547136  1.785020  0.020165   

      conviction  zhangs_metric  
41      1.043831       0.674500  
223     1.661306       0.733956  
349     1.029031       0.396595  
461     1.096954       0.776559  
621     1.498210       0.706393  
...          ...            ...  
5045    1.092634       0.746171  
5047    1.201652       0.706745  
5049    1.470284       0.768804  
5051    1.046733       0.551025  
5053    1.531331       0.480009  

[82 rows x 10 columns]

In [ ]:
#se ordena segun soporte
mayonesa_soporte = mayonesa.sort_values('support', ascending= False)
mayonesa_soporte.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
5053  frozenset({'MAYONESA'})           (YOGHURT)            0.083804   
5027  frozenset({'MAYONESA'})  (SALSAS DE TOMATE)            0.083804   
223   frozenset({'MAYONESA'})    (ACEITE VEGETAL)            0.083804   
1101  frozenset({'MAYONESA'})  (AZUCAR GRANULADA)            0.083804   
3309  frozenset({'MAYONESA'})       (FIDEO CORTO)            0.083804   
3451  frozenset({'MAYONESA'})       (FIDEO LARGO)            0.083804   
4548  frozenset({'MAYONESA'})   (LECHES LIQUIDAS)            0.083804   
621   frozenset({'MAYONESA'})     (ARROZ REGULAR)            0.083804   
3649  frozenset({'MAYONESA'})   (GALLETAS DULCES)            0.083804   
4919  frozenset({'MAYONESA'})         (MARGARINA)            0.083804   

      consequent support   support  confidence      lift  leverage  \
5053            0.306515  0.045852    0.547136  1.785020  0.020165   
5027            0.171559  0.043952    0.524463  3.057049  0.029575   
223             0.162408  0.041552    0.495823  3.052947  0.027942   
1101            0.180459  0.040352    0.481504  2.668216  0.025229   
3309            0.159108  0.039052    0.465990  2.928769  0.025718   
3451            0.154708  0.038502    0.459427  2.969646  0.025537   
4548            0.231312  0.036852    0.439737  1.901061  0.017467   
621             0.153458  0.036452    0.434964  2.834425  0.023591   
3649            0.255163  0.034652    0.413484  1.620473  0.013268   
4919            0.141807  0.034002    0.405728  2.861126  0.022118   

      conviction  zhangs_metric  
5053    1.531331       0.480009  
5027    1.742118       0.734436  
223     1.661306       0.733956  
1101    1.580611       0.682406  
3309    1.574676       0.718798  
3451    1.563697       0.723928  
4548    1.372015       0.517333  
621     1.498210       0.706393  
3649    1.269936       0.417920  
4919    1.444108       0.709987

In [ ]:
#se ordena segun confianza
mayonesa_confianza = mayonesa.sort_values('confidence', ascending= False)
mayonesa_confianza.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
5053  frozenset({'MAYONESA'})           (YOGHURT)            0.083804   
5027  frozenset({'MAYONESA'})  (SALSAS DE TOMATE)            0.083804   
223   frozenset({'MAYONESA'})    (ACEITE VEGETAL)            0.083804   
1101  frozenset({'MAYONESA'})  (AZUCAR GRANULADA)            0.083804   
3309  frozenset({'MAYONESA'})       (FIDEO CORTO)            0.083804   
3451  frozenset({'MAYONESA'})       (FIDEO LARGO)            0.083804   
4548  frozenset({'MAYONESA'})   (LECHES LIQUIDAS)            0.083804   
621   frozenset({'MAYONESA'})     (ARROZ REGULAR)            0.083804   
3649  frozenset({'MAYONESA'})   (GALLETAS DULCES)            0.083804   
4919  frozenset({'MAYONESA'})         (MARGARINA)            0.083804   

      consequent support   support  confidence      lift  leverage  \
5053            0.306515  0.045852    0.547136  1.785020  0.020165   
5027            0.171559  0.043952    0.524463  3.057049  0.029575   
223             0.162408  0.041552    0.495823  3.052947  0.027942   
1101            0.180459  0.040352    0.481504  2.668216  0.025229   
3309            0.159108  0.039052    0.465990  2.928769  0.025718   
3451            0.154708  0.038502    0.459427  2.969646  0.025537   
4548            0.231312  0.036852    0.439737  1.901061  0.017467   
621             0.153458  0.036452    0.434964  2.834425  0.023591   
3649            0.255163  0.034652    0.413484  1.620473  0.013268   
4919            0.141807  0.034002    0.405728  2.861126  0.022118   

      conviction  zhangs_metric  
5053    1.531331       0.480009  
5027    1.742118       0.734436  
223     1.661306       0.733956  
1101    1.580611       0.682406  
3309    1.574676       0.718798  
3451    1.563697       0.723928  
4548    1.372015       0.517333  
621     1.498210       0.706393  
3649    1.269936       0.417920  
4919    1.444108       0.709987

In [ ]:
#se ordena segun ganancia
mayonesa_ganancia = mayonesa.sort_values('lift', ascending= False)
mayonesa_ganancia.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents           consequents  antecedent support  \
4353  frozenset({'MAYONESA'})             (KETCHUP)            0.083804   
5001  frozenset({'MAYONESA'})             (MOSTAZA)            0.083804   
5037  frozenset({'MAYONESA'})  (SUCEDANEO DE LIMON)            0.083804   
2831  frozenset({'MAYONESA'})              (CREMAS)            0.083804   
5013  frozenset({'MAYONESA'})        (PASTA DENTAL)            0.083804   

      consequent support   support  confidence      lift  leverage  \
4353            0.043002  0.027251    0.325179  7.561924  0.023648   
5001            0.020901  0.012201    0.145585  6.965428  0.010449   
5037            0.016851  0.005950    0.071002  4.213581  0.004538   
2831            0.020951  0.006800    0.081146  3.873104  0.005045   
5013            0.024751  0.007950    0.094869  3.832889  0.005876   

      conviction  zhangs_metric  
4353    1.418151       0.947132  
5001    1.145929       0.934772  
5037    1.058290       0.832434  
2831    1.065510       0.809662  
5013    1.077467       0.806705

## Galletas dulces

In [ ]:
galletas = rules_100[rules_100['antecedents'].str.contains('GALLETAS DULCES')]
galletas

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents            consequents  \
23    frozenset({'GALLETAS DULCES'})  (ACEITE DE MARAVILLA)   
81    frozenset({'GALLETAS DULCES'})      (ACEITE DE OLIVA)   
168   frozenset({'GALLETAS DULCES'})       (ACEITE VEGETAL)   
332   frozenset({'GALLETAS DULCES'})                (AGUAS)   
380   frozenset({'GALLETAS DULCES'})              (ALFAJOR)   
...                              ...                    ...   
3725  frozenset({'GALLETAS DULCES'})             (VIENESAS)   
3726  frozenset({'GALLETAS DULCES'})             (VINAGRES)   
3729  frozenset({'GALLETAS DULCES'})                (VINOS)   
3730  frozenset({'GALLETAS DULCES'})    (YERBA MATE GRANEL)   
3732  frozenset({'GALLETAS DULCES'})              (YOGHURT)   

      antecedent support  consequent support   support  confidence      lift  \
23              0.255163            0.025301  0.012101    0.047423  1.874336   
81              0.255163            0.012001  0.005650    0.022144  1.845227   
168             0.255163            0.162408  0.059553    0.233392  1.437072   
332             0.255163            0.047102  0.018301    0.071723  1.522695   
380             0.255163            0.007900  0.005100    0.019988  2.530031   
...                  ...                 ...       ...         ...       ...   
3725            0.255163            0.118356  0.047952    0.187929  1.587827   
3726            0.255163            0.014051  0.005850    0.022928  1.631782   
3729            0.255163            0.041952  0.012851    0.050363  1.200477   
3730            0.255163            0.012701  0.005350    0.020968  1.650946   
3732            0.255163            0.306515  0.133507    0.523222  1.707000   

      leverage  conviction  zhangs_metric  
23    0.005645    1.023223       0.626282  
81    0.002588    1.010373       0.614982  
168   0.018112    1.092595       0.408332  
332   0.006282    1.026522       0.460865  
380   0.003084    1.012334       0.811920  
...        ...         ...            ...  
3725  0.017752    1.085673       0.497032  
3726  0.002265    1.009085       0.519809  
3729  0.002146    1.008856       0.224207  
3730  0.002110    1.008444       0.529359  
3732  0.055295    1.454522       0.556064  

[119 rows x 10 columns]

In [ ]:
galletas_soporte = galletas.sort_values('support', ascending= False)
galletas_soporte.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
3732  frozenset({'GALLETAS DULCES'})            (YOGHURT)            0.255163   
3628  frozenset({'GALLETAS DULCES'})    (LECHES LIQUIDAS)            0.255163   
3657  frozenset({'GALLETAS DULCES'})  (NECTARES LIQUIDOS)            0.255163   
3698  frozenset({'GALLETAS DULCES'})   (SALSAS DE TOMATE)            0.255163   
1040  frozenset({'GALLETAS DULCES'})   (AZUCAR GRANULADA)            0.255163   

      consequent support   support  confidence      lift  leverage  \
3732            0.306515  0.133507    0.523222  1.707000  0.055295   
3628            0.231312  0.105455    0.413286  1.786708  0.046433   
3657            0.178859  0.090655    0.355281  1.986377  0.045016   
3698            0.171559  0.071004    0.278268  1.621998  0.027228   
1040            0.180459  0.066753    0.261611  1.449697  0.020707   

      conviction  zhangs_metric  
3732    1.454522       0.556064  
3628    1.310159       0.591151  
3657    1.273642       0.666684  
3698    1.147851       0.514846  
1040    1.109904       0.416467

In [ ]:
galletas_confianza = galletas.sort_values('confidence', ascending= False)
galletas_confianza.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  antecedent support  \
3732  frozenset({'GALLETAS DULCES'})            (YOGHURT)            0.255163   
3628  frozenset({'GALLETAS DULCES'})    (LECHES LIQUIDAS)            0.255163   
3657  frozenset({'GALLETAS DULCES'})  (NECTARES LIQUIDOS)            0.255163   
3698  frozenset({'GALLETAS DULCES'})   (SALSAS DE TOMATE)            0.255163   
1040  frozenset({'GALLETAS DULCES'})   (AZUCAR GRANULADA)            0.255163   

      consequent support   support  confidence      lift  leverage  \
3732            0.306515  0.133507    0.523222  1.707000  0.055295   
3628            0.231312  0.105455    0.413286  1.786708  0.046433   
3657            0.178859  0.090655    0.355281  1.986377  0.045016   
3698            0.171559  0.071004    0.278268  1.621998  0.027228   
1040            0.180459  0.066753    0.261611  1.449697  0.020707   

      conviction  zhangs_metric  
3732    1.454522       0.556064  
3628    1.310159       0.591151  
3657    1.273642       0.666684  
3698    1.147851       0.514846  
1040    1.109904       0.416467

In [ ]:
galletas_gannacia = galletas.sort_values('lift', ascending= False)
galletas_gannacia.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents        consequents  antecedent support  \
3658  frozenset({'GALLETAS DULCES'})   (OBLEAS BANADAS)            0.255163   
1507  frozenset({'GALLETAS DULCES'})        (BIZCOCHOS)            0.255163   
380   frozenset({'GALLETAS DULCES'})          (ALFAJOR)            0.255163   
2190  frozenset({'GALLETAS DULCES'})  (CERAL AZUCARADO)            0.255163   
2214  frozenset({'GALLETAS DULCES'})  (CERAL EN BARRAS)            0.255163   

      consequent support   support  confidence      lift  leverage  \
3658            0.016751  0.011201    0.043896  2.620511  0.006926   
1507            0.047902  0.031052    0.121693  2.540439  0.018829   
380             0.007900  0.005100    0.019988  2.530031  0.003084   
2190            0.016051  0.009950    0.038997  2.429577  0.005855   
2214            0.028551  0.017651    0.069175  2.422821  0.010366   

      conviction  zhangs_metric  
3658    1.028391       0.830242  
1507    1.084015       0.814094  
380     1.012334       0.811920  
2190    1.023877       0.789979  
2214    1.043643       0.788438

## Verduras congeladas

In [ ]:
verduras = rules_100[rules_100['antecedents'].str.contains('VERDURAS CONGELADAS')]
verduras

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents            consequents  \
67    frozenset({'VERDURAS CONGELADAS'})  (ACEITE DE MARAVILLA)   
296   frozenset({'VERDURAS CONGELADAS'})       (ACEITE VEGETAL)   
374   frozenset({'VERDURAS CONGELADAS'})                (AGUAS)   
491   frozenset({'VERDURAS CONGELADAS'})    (ARROZ PREGRANEADO)   
692   frozenset({'VERDURAS CONGELADAS'})        (ARROZ REGULAR)   
...                                  ...                    ...   
5782  frozenset({'VERDURAS CONGELADAS'})     (TOALLAS DE PAPEL)   
5791  frozenset({'VERDURAS CONGELADAS'})   (TOALLAS HIGIENICAS)   
5796  frozenset({'VERDURAS CONGELADAS'})             (VIENESAS)   
5799  frozenset({'VERDURAS CONGELADAS'})                (VINOS)   
5800  frozenset({'VERDURAS CONGELADAS'})              (YOGHURT)   

      antecedent support  consequent support   support  confidence      lift  \
67              0.083704            0.025301  0.006400    0.076464  3.022124   
296             0.083704            0.162408  0.033252    0.397252  2.446011   
374             0.083704            0.047102  0.007950    0.094982  2.016504   
491             0.083704            0.034602  0.010051    0.120072  3.470106   
692             0.083704            0.153458  0.035102    0.419355  2.732707   
...                  ...                 ...       ...         ...       ...   
5782            0.083704            0.106105  0.022751    0.271804  2.561644   
5791            0.083704            0.037752  0.009750    0.116487  3.085606   
5796            0.083704            0.118356  0.030952    0.369773  3.124246   
5799            0.083704            0.041952  0.006500    0.077658  1.851118   
5800            0.083704            0.306515  0.048302    0.577061  1.882650   

      leverage  conviction  zhangs_metric  
67    0.004282    1.055398       0.730230  
296   0.019657    1.389622       0.645175  
374   0.004008    1.052905       0.550141  
491   0.007154    1.097133       0.776850  
692   0.022257    1.457934       0.691984  
...        ...         ...            ...  
5782  0.013870    1.227547       0.665315  
5791  0.006590    1.089117       0.737660  
5796  0.021045    1.398931       0.742034  
5799  0.002989    1.038713       0.501788  
5800  0.022646    1.639680       0.511662  

[80 rows x 10 columns]

In [ ]:
verduras_soporte = verduras.sort_values('support', ascending= False)
verduras_soporte.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  \
5800  frozenset({'VERDURAS CONGELADAS'})           (YOGHURT)   
5646  frozenset({'VERDURAS CONGELADAS'})  (SALSAS DE TOMATE)   
3385  frozenset({'VERDURAS CONGELADAS'})       (FIDEO CORTO)   
4630  frozenset({'VERDURAS CONGELADAS'})   (LECHES LIQUIDAS)   
3722  frozenset({'VERDURAS CONGELADAS'})   (GALLETAS DULCES)   

      antecedent support  consequent support   support  confidence      lift  \
5800            0.083704            0.306515  0.048302    0.577061  1.882650   
5646            0.083704            0.171559  0.040152    0.479689  2.796068   
3385            0.083704            0.159108  0.039402    0.470729  2.958550   
4630            0.083704            0.231312  0.038902    0.464755  2.009217   
3722            0.083704            0.255163  0.036402    0.434886  1.704349   

      leverage  conviction  zhangs_metric  
5800  0.022646    1.639680       0.511662  
5646  0.025792    1.592205       0.701034  
3385  0.026084    1.588773       0.722470  
4630  0.019540    1.436143       0.548178  
3722  0.015044    1.318031       0.451018

In [ ]:
verduras_confianza = verduras.sort_values('confidence', ascending= False)
verduras_confianza.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  \
5800  frozenset({'VERDURAS CONGELADAS'})           (YOGHURT)   
5646  frozenset({'VERDURAS CONGELADAS'})  (SALSAS DE TOMATE)   
3385  frozenset({'VERDURAS CONGELADAS'})       (FIDEO CORTO)   
4630  frozenset({'VERDURAS CONGELADAS'})   (LECHES LIQUIDAS)   
3722  frozenset({'VERDURAS CONGELADAS'})   (GALLETAS DULCES)   

      antecedent support  consequent support   support  confidence      lift  \
5800            0.083704            0.306515  0.048302    0.577061  1.882650   
5646            0.083704            0.171559  0.040152    0.479689  2.796068   
3385            0.083704            0.159108  0.039402    0.470729  2.958550   
4630            0.083704            0.231312  0.038902    0.464755  2.009217   
3722            0.083704            0.255163  0.036402    0.434886  1.704349   

      leverage  conviction  zhangs_metric  
5800  0.022646    1.639680       0.511662  
5646  0.025792    1.592205       0.701034  
3385  0.026084    1.588773       0.722470  
4630  0.019540    1.436143       0.548178  
3722  0.015044    1.318031       0.451018

In [ ]:
verduras_ganancia = verduras.sort_values('lift', ascending= False)
verduras_ganancia.head(5)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents          consequents  \
2170  frozenset({'VERDURAS CONGELADAS'})  (CARNES CONGELADAS)   
746   frozenset({'VERDURAS CONGELADAS'})            (ARVEJAS)   
5228  frozenset({'VERDURAS CONGELADAS'})            (PALMITO)   
2659  frozenset({'VERDURAS CONGELADAS'})            (CHORIZO)   
3901  frozenset({'VERDURAS CONGELADAS'})       (HAMBURGUESAS)   

      antecedent support  consequent support   support  confidence      lift  \
2170            0.083704            0.037452  0.014601    0.174432  4.657511   
746             0.083704            0.020601  0.007800    0.093190  4.523558   
5228            0.083704            0.015951  0.005550    0.066308  4.157049   
2659            0.083704            0.016701  0.005600    0.066906  4.006124   
3901            0.083704            0.079454  0.025951    0.310036  3.902081   

      leverage  conviction  zhangs_metric  
2170  0.011466    1.165923       0.857030  
746   0.006076    1.080049       0.850091  
5228  0.004215    1.053934       0.828820  
2659  0.004202    1.053805       0.818930  
3901  0.019301    1.334194       0.811666

## Salsas de tomate

In [ ]:
salsastomate = rules_100[rules_100['antecedents'].str.contains('SALSAS DE TOMATE')]
salsastomate

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents             consequents  \
57    frozenset({'SALSAS DE TOMATE'})   (ACEITE DE MARAVILLA)   
87    frozenset({'SALSAS DE TOMATE'})       (ACEITE DE OLIVA)   
268   frozenset({'SALSAS DE TOMATE'})        (ACEITE VEGETAL)   
364   frozenset({'SALSAS DE TOMATE'})                 (AGUAS)   
399   frozenset({'SALSAS DE TOMATE'})  (ALIMENTO PARA PERROS)   
...                               ...                     ...   
5649  frozenset({'SALSAS DE TOMATE'})              (VIENESAS)   
5650  frozenset({'SALSAS DE TOMATE'})              (VINAGRES)   
5653  frozenset({'SALSAS DE TOMATE'})                 (VINOS)   
5654  frozenset({'SALSAS DE TOMATE'})     (YERBA MATE GRANEL)   
5656  frozenset({'SALSAS DE TOMATE'})               (YOGHURT)   

      antecedent support  consequent support   support  confidence      lift  \
57              0.171559            0.025301  0.013601    0.079277  3.133329   
87              0.171559            0.012001  0.006200    0.036141  3.011605   
268             0.171559            0.162408  0.091355    0.532498  3.278763   
364             0.171559            0.047102  0.011501    0.067036  1.423195   
399             0.171559            0.020951  0.006800    0.039639  1.891962   
...                  ...                 ...       ...         ...       ...   
5649            0.171559            0.118356  0.059903    0.349169  2.950164   
5650            0.171559            0.014051  0.008300    0.048382  3.443415   
5653            0.171559            0.041952  0.010951    0.063830  1.521492   
5654            0.171559            0.012701  0.005900    0.034392  2.707920   
5656            0.171559            0.306515  0.098405    0.573594  1.871338   

      leverage  conviction  zhangs_metric  
57    0.009260    1.058623       0.821845  
87    0.004142    1.025046       0.806274  
268   0.063492    1.791632       0.838933  
364   0.003420    1.021366       0.358934  
399   0.003206    1.019459       0.569079  
...        ...         ...            ...  
5649  0.039598    1.354644       0.797927  
5650  0.005890    1.036077       0.856537  
5653  0.003753    1.023369       0.413729  
5654  0.003721    1.022464       0.761325  
5656  0.045820    1.626347       0.562047  

[113 rows x 10 columns]

In [ ]:
salsastomate_soporte = salsastomate.sort_values('support', ascending= False)
salsastomate_soporte.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
3357  frozenset({'SALSAS DE TOMATE'})       (FIDEO CORTO)            0.171559   
3495  frozenset({'SALSAS DE TOMATE'})       (FIDEO LARGO)            0.171559   
5656  frozenset({'SALSAS DE TOMATE'})           (YOGHURT)            0.171559   
1150  frozenset({'SALSAS DE TOMATE'})  (AZUCAR GRANULADA)            0.171559   
666   frozenset({'SALSAS DE TOMATE'})     (ARROZ REGULAR)            0.171559   
268   frozenset({'SALSAS DE TOMATE'})    (ACEITE VEGETAL)            0.171559   
4602  frozenset({'SALSAS DE TOMATE'})   (LECHES LIQUIDAS)            0.171559   
3699  frozenset({'SALSAS DE TOMATE'})   (GALLETAS DULCES)            0.171559   
5347  frozenset({'SALSAS DE TOMATE'})   (PAPEL HIGIENICO)            0.171559   
4959  frozenset({'SALSAS DE TOMATE'})         (MARGARINA)            0.171559   

      consequent support   support  confidence      lift  leverage  \
3357            0.159108  0.107055    0.624016  3.921968  0.079759   
3495            0.154708  0.106505    0.620810  4.012794  0.079964   
5656            0.306515  0.098405    0.573594  1.871338  0.045820   
1150            0.180459  0.095805    0.558438  3.094541  0.064845   
666             0.153458  0.094355    0.549985  3.583955  0.068028   
268             0.162408  0.091355    0.532498  3.278763  0.063492   
4602            0.231312  0.078304    0.456427  1.973212  0.038620   
3699            0.255163  0.071004    0.413874  1.621998  0.027228   
5347            0.160508  0.067303    0.392305  2.444149  0.039767   
4959            0.141807  0.066153    0.385602  2.719200  0.041825   

      conviction  zhangs_metric  
3357    2.236512       0.899310  
3495    2.229207       0.906277  
5656    1.626347       0.562047  
1150    1.856003       0.817016  
666     1.881144       0.870283  
268     1.791632       0.838933  
4602    1.414139       0.595349  
3699    1.270779       0.462889  
5347    1.381437       0.713218  
4959    1.396803       0.763174

In [ ]:
salsastomate_confianza = salsastomate.sort_values('confidence', ascending= False)
salsastomate_confianza.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents         consequents  antecedent support  \
3357  frozenset({'SALSAS DE TOMATE'})       (FIDEO CORTO)            0.171559   
3495  frozenset({'SALSAS DE TOMATE'})       (FIDEO LARGO)            0.171559   
5656  frozenset({'SALSAS DE TOMATE'})           (YOGHURT)            0.171559   
1150  frozenset({'SALSAS DE TOMATE'})  (AZUCAR GRANULADA)            0.171559   
666   frozenset({'SALSAS DE TOMATE'})     (ARROZ REGULAR)            0.171559   
268   frozenset({'SALSAS DE TOMATE'})    (ACEITE VEGETAL)            0.171559   
4602  frozenset({'SALSAS DE TOMATE'})   (LECHES LIQUIDAS)            0.171559   
3699  frozenset({'SALSAS DE TOMATE'})   (GALLETAS DULCES)            0.171559   
5347  frozenset({'SALSAS DE TOMATE'})   (PAPEL HIGIENICO)            0.171559   
4959  frozenset({'SALSAS DE TOMATE'})         (MARGARINA)            0.171559   

      consequent support   support  confidence      lift  leverage  \
3357            0.159108  0.107055    0.624016  3.921968  0.079759   
3495            0.154708  0.106505    0.620810  4.012794  0.079964   
5656            0.306515  0.098405    0.573594  1.871338  0.045820   
1150            0.180459  0.095805    0.558438  3.094541  0.064845   
666             0.153458  0.094355    0.549985  3.583955  0.068028   
268             0.162408  0.091355    0.532498  3.278763  0.063492   
4602            0.231312  0.078304    0.456427  1.973212  0.038620   
3699            0.255163  0.071004    0.413874  1.621998  0.027228   
5347            0.160508  0.067303    0.392305  2.444149  0.039767   
4959            0.141807  0.066153    0.385602  2.719200  0.041825   

      conviction  zhangs_metric  
3357    2.236512       0.899310  
3495    2.229207       0.906277  
5656    1.626347       0.562047  
1150    1.856003       0.817016  
666     1.881144       0.870283  
268     1.791632       0.838933  
4602    1.414139       0.595349  
3699    1.270779       0.462889  
5347    1.381437       0.713218  
4959    1.396803       0.763174

In [ ]:
salsastomate_ganancia = salsastomate.sort_values('lift', ascending= False)
salsastomate_ganancia.head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents           consequents  \
4404  frozenset({'SALSAS DE TOMATE'})             (LASAGNA)   
4664  frozenset({'SALSAS DE TOMATE'})            (LENTEJAS)   
3495  frozenset({'SALSAS DE TOMATE'})         (FIDEO LARGO)   
5468  frozenset({'SALSAS DE TOMATE'})             (POROTOS)   
3357  frozenset({'SALSAS DE TOMATE'})         (FIDEO CORTO)   
2844  frozenset({'SALSAS DE TOMATE'})              (CREMAS)   
479   frozenset({'SALSAS DE TOMATE'})   (ARROZ PREGRANEADO)   
4317  frozenset({'SALSAS DE TOMATE'})               (JUREL)   
5624  frozenset({'SALSAS DE TOMATE'})             (SHAMPOO)   
5631  frozenset({'SALSAS DE TOMATE'})  (SUCEDANEO DE LIMON)   

      antecedent support  consequent support   support  confidence      lift  \
4404            0.171559            0.012701  0.009650    0.056252  4.429056   
4664            0.171559            0.022451  0.015601    0.090936  4.050380   
3495            0.171559            0.154708  0.106505    0.620810  4.012794   
5468            0.171559            0.013251  0.009100    0.053046  4.003253   
3357            0.171559            0.159108  0.107055    0.624016  3.921968   
2844            0.171559            0.020951  0.014001    0.081609  3.895216   
479             0.171559            0.034602  0.022101    0.128825  3.723092   
4317            0.171559            0.068553  0.043052    0.250947  3.660608   
5624            0.171559            0.028951  0.018101    0.105509  3.644329   
5631            0.171559            0.016851  0.010451    0.060915  3.614964   

      leverage  conviction  zhangs_metric  
4404  0.007472    1.046147       0.934548  
4664  0.011749    1.075335       0.909068  
3495  0.079964    2.229207       0.906277  
5468  0.006827    1.042024       0.905560  
3357  0.079759    2.236512       0.899310  
2844  0.010406    1.066048       0.897197  
479   0.016165    1.108157       0.882870  
4317  0.031291    1.243499       0.877336  
5624  0.013134    1.085587       0.875863  
5631  0.007560    1.046923       0.873172

# Pregunta 4
#### Defina tres categorías a recomendar para cada una de las 10 primeras transacciones de la BD. Explique qué criterios usaría para priorizar qué regla o categoría recomendar.



Helper function. Pon un array de categorías y te da el margen ordenado de forma descendiente

In [ ]:
#se muestran las primeras 10 transacciones
data.iloc[:, 2:].head(10)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


ACEITE DE MARAVILLA  ACEITE DE OLIVA  ACEITE PEPITA DE UVA  ACEITE VEGETAL  \
0                    0                0                     0               0   
1                    0                0                     0               1   
2                    0                0                     0               0   
3                    0                0                     0               1   
4                    0                0                     0               0   
5                    0                0                     0               0   
6                    0                0                     0               0   
7                    0                0                     0               1   
8                    0                0                     0               0   
9                    0                0                     0               0   

   ACEITUNAS  ADHESIVOS  AGUARDIENTE  AGUAS  AJI  AJO  ...  VERDURAS  \
0          0          0            0      0    0    0  ...         0   
1          0          0            0      0    0    0  ...         0   
2          0          0            0      0    0    0  ...         0   
3          0          0            0      0    0    0  ...         0   
4          0          0            0      0    0    0  ...         0   
5          0          0            0      0    0    0  ...         0   
6          0          0            0      0    0    0  ...         0   
7          0          0            0      0    0    0  ...         0   
8          0          0            0      0    0    0  ...         0   
9          0          0            0      0    0    0  ...         0   

   VERDURAS CONGELADAS  VERMOUTH  VIENESAS  VINAGRES  VINOS  VIRUTILLAS  \
0                    0         0         0         0      0           0   
1                    0         0         0         0      0           0   
2                    1         0         0         0      0           0   
3                    0         0         0         0      0           0   
4                    0         0         0         0      0           0   
5                    0         0         0         0      0           0   
6                    0         0         0         0      1           0   
7                    1         0         1         0      0           0   
8                    0         0         0         0      0           0   
9                    0         0         0         0      0           0   

   WHISKY  YERBA MATE GRANEL  YOGHURT  
0       0                  0        0  
1       0                  0        0  
2       0                  0        0  
3       0                  0        1  
4       0                  0        0  
5       0                  0        0  
6       0                  0        0  
7       0                  0        1  
8       0                  1        0  
9       0                  0        0  

[10 rows x 277 columns]

### Compra 1

In [ ]:
#se considera la primer fila de la tabla
compra1 = data.iloc[0,2:]

#se imprimen los productos comprados
productos = compra1[compra1 == 1]
print("\nProductos comprados: ", productos)

#se filtra la tabla de reglas para que muestre sólo los antecedentes deseados
durazno = rules_100[rules_100['antecedents'].str.contains('DURAZNO EN CONSERVA')]


Productos comprados:  DURAZNO EN CONSERVA    1
Name: 0, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
import plotly.express as px
import pandas as pd

#se grafican los valores de las reglas encontradas
fig = px.scatter(durazno, x="support", y="confidence", color="lift",
                 hover_name=durazno['antecedents'].astype(str) + ' -> ' + durazno['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


### Compra 2

In [ ]:
compra2 = data.iloc[1,2:]

productos = compra2[compra2 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  ACEITE VEGETAL          1
AZUCAR GRANULADA        1
JUGOS EN POLVO          1
LIQUIDO PARA PISOS      1
PANOS Y FIBRAS SINTE    1
Name: 1, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra2_sum = rules_100[rules_100['antecedents'].str.contains('ACEITE VEGETAL|AZUCAR GRANULADA|JUGOS EN POLVO|LIQUIDO PARA PISOS|PANOS Y FIBRAS SINTE')]
fig = px.scatter(compra2_sum, x="support", y="confidence", color="lift",
                 hover_name=compra2_sum['antecedents'].astype(str) + ' -> ' + compra2_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 3

In [ ]:
compra3 = data.iloc[2,2:]

productos = compra3[compra3 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  GALLETAS DULCES        1
JUGOS EN POLVO         1
LECHES LIQUIDAS        1
OTRAS CECINAS          1
PAPEL HIGIENICO        1
PATE                   1
QUESOS                 1
TOALLAS DE PAPEL       1
VERDURAS CONGELADAS    1
Name: 2, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra3_sum = rules_100[rules_100['antecedents'].str.contains('GALLETAS DULCES|LECHES LIQUIDAS|JUGOS EN POLVO|OTRAS CECINAS|PAPEL HIGIENICO|PATE|QUESOS|TOALLAS DE PAPEL|VERDURAS CONGELADAS')]
fig = px.scatter(compra3_sum, x="support", y="confidence", color="lift",
                 hover_name=compra3_sum['antecedents'].astype(str) + ' -> ' + compra3_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 4

In [ ]:
compra4 = data.iloc[3,2:]

productos = compra4[compra4 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  ACEITE VEGETAL          1
AMPOLLETAS              1
ARROZ PREGRANEADO       1
ARROZ REGULAR           1
ATUN                    1
AZUCAR GRANULADA        1
CAFE INSTAN PURO REG    1
CEREALES                1
CLOROS LIQUIDOS         1
FIDEO CORTO             1
FIDEO LARGO             1
FOSFOROS                1
HARINA ENVASADA         1
LENTEJAS                1
LIQUIDO PARA PISOS      1
MAIZENA                 1
MARGARINA               1
PAPEL ALUSA             1
SALSAS DE TOMATE        1
SOPAS                   1
TE EN BOLSA             1
TE EN GRANEL            1
TOALLAS HIGIENICAS      1
YOGHURT                 1
Name: 3, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra4_sum = rules_100[rules_100['antecedents'].str.contains('ACEITE VEGETAL|AMPOLLETAS|ARROZ PREGRANEADO|ARROZ REGULAR|ATUN|AZUCAR GRANULADA|CAFE INSTAN PURO REG|CEREALES|CLOROS LIQUIDOS|FIDEO CORTO|FIDEO LARGO|FOSFOROS|HARINA ENVASADA|LENTEJAS|LIQUIDO PARA PISOS|MAIZENA|MARGARINA|PAPEL ALUSA|SALSAS DE TOMATE|SOPAS|TE EN BOLSA|TE EN GRANEL|TOALLAS HIGIENICAS|YOGHURT')]
fig = px.scatter(compra4_sum, x="support", y="confidence", color="lift",
                 hover_name=compra4_sum['antecedents'].astype(str) + ' -> ' + compra4_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 5

In [ ]:
compra5 = data.iloc[4,2:]

productos = compra5[compra5 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  BEBIDAS GASEOSAS    1
CARAMELOS DUROS     1
GALLETAS DULCES     1
PAPAS FRITAS        1
RAMITAS             1
SIN DESCRIPCION     1
SUFLE               1
Name: 4, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra5_sum = rules_100[rules_100['antecedents'].str.contains('BEBIDAS GASEOSAS|CARAMELOS DUROS|GALLETAS DULCES|PAPAS FRITAS|RAMITAS|SIN DESCRIPCION|SUFLE')]
fig = px.scatter(compra5_sum, x="support", y="confidence", color="lift",
                 hover_name=compra5_sum['antecedents'].astype(str) + ' -> ' + compra5_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 6

In [ ]:
compra6 = data.iloc[5,2:]

productos = compra6[compra6 == 1]
print("\nProductos comprados: ", productos)

categorias = ['RAMITAS', 'SUFLE', 'SURTIDOS DE SNACKS']
print_margen_by_categories(categorias)


Productos comprados:  AZUCAR GRANULADA        1
BEBIDA LACTEA           1
BEBIDAS GASEOSAS        1
CARAMELOS BLANDOS       1
CERAL AZUCARADO         1
CHOCOLATES RELLENOS     1
GALLETAS DULCES         1
HELADOS                 1
NECTARES LIQUIDOS       1
OTROS CONFITES          1
POSTRES REFRIGERADOS    1
SIN DESCRIPCION         1
SURTIDO FRUT CONSER     1
Name: 5, dtype: int64
              Categoria  Margen promedio
253  SURTIDOS DE SNACKS              834
249               SUFLE              787
231             RAMITAS              375


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra6_sum = rules_100[rules_100['antecedents'].str.contains('AZUCAR GRANULADA|BEBIDA LACTEA|BEBIDAS GASEOSAS|CARAMELOS BLANDOS|CERAL AZUCARADO|CHOCOLATES RELLENOS|GALLETAS DULCES|HELADOS|NECTARES LIQUIDOS|OTROS CONFITES|POSTRES REFRIGERADOS|SIN DESCRIPCION|SURTIDO FRUT CONSER')]
fig = px.scatter(compra6_sum, x="support", y="confidence", color="lift",
                 hover_name=compra6_sum['antecedents'].astype(str) + ' -> ' + compra6_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 7

In [ ]:
compra7 = data.iloc[6,2:]

productos = compra7[compra7 == 1]
print("\nProductos comprados: ", productos)

categorias = ['PISCO', 'COMIDAS LISTAS', 'CREMAS']
print_margen_by_categories(categorias)


Productos comprados:  BEBIDAS GASEOSAS    1
GALLETAS DULCES     1
LECHE CONDENSADA    1
LECHES LIQUIDAS     1
MARGARINA           1
SOPAS               1
VINOS               1
Name: 6, dtype: int64
          Categoria  Margen promedio
80   COMIDAS LISTAS              674
219           PISCO              339
85           CREMAS              276


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra7_sum = rules_100[rules_100['antecedents'].str.contains('BEBIDAS GASEOSAS|GALLETAS DULCES|LECHE CONDENSADA|LECHES LIQUIDAS|MARGARINA|SOPAS|VINOS')]
fig = px.scatter(compra7_sum, x="support", y="confidence", color="lift",
                 hover_name=compra7_sum['antecedents'].astype(str) + ' -> ' + compra7_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 8

In [ ]:
compra8 = data.iloc[7,2:]

productos = compra8[compra8 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  ACEITE VEGETAL         1
ARROZ REGULAR          1
CLOROS LIQUIDOS        1
FIDEO CORTO            1
FIDEO LARGO            1
GALLETAS DULCES        1
HAMBURGUESAS           1
LAVALOZAS              1
LECHES LIQUIDAS        1
LIQUIDO PARA PISOS     1
MANJAR                 1
PASTA DENTAL           1
SALSAS DE TOMATE       1
TOALLAS HIGIENICAS     1
VERDURAS CONGELADAS    1
VIENESAS               1
YOGHURT                1
Name: 7, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra8_sum = rules_100[rules_100['antecedents'].str.contains('ACEITE VEGETAL|ARROZ REGULAR|CLOROS LIQUIDOS|FIDEO CORTO|FIDEO LARGO|GALLETAS DULCES|HAMBURGUESAS|LAVALOZAS|LECHES LIQUIDAS|LIQUIDO PARA PISOS|MANJAR|PASTA DENTAL|SALSAS DE TOMATE|TOALLAS HIGIENICAS|VERDURAS CONGELADAS|VIENESAS|YOGHURT')]
fig = px.scatter(compra8_sum, x="support", y="confidence", color="lift",
                 hover_name=compra8_sum['antecedents'].astype(str) + ' -> ' + compra8_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



### Compra 9

In [ ]:
compra9 = data.iloc[8,2:]

productos = compra9[compra9 == 1]
print("\nProductos comprados: ", productos)


Productos comprados:  ALFAJOR                 1
AZUCAR GRANULADA        1
BEBIDAS GASEOSAS        1
CAFE INSTAN PURO REG    1
DURAZNO EN CONSERVA     1
JUREL                   1
KETCHUP                 1
LEVADURA GRANULADA      1
MANJAR                  1
MAYONESA                1
TE EN BOLSA             1
YERBA MATE GRANEL       1
Name: 8, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
compra9_sum = rules_100[rules_100['antecedents'].str.contains('ALFAJOR|AZUCAR GRANULADA|BEBIDAS GASEOSAS|CAFE INSTAN PURO REG|DURAZNO EN CONSERVA |JUREL|KETCHUP|LEVADURA GRANULADA|MANJAR|MAYONESA|TE EN BOLSA|YERBA MATE GRANEL')]
fig = px.scatter(compra9_sum, x="support", y="confidence", color="lift",
                 hover_name=compra9_sum['antecedents'].astype(str) + ' -> ' + compra9_sum['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



Compra 10

In [ ]:
compra10 = data.iloc[9,2:]

productos = compra10[compra10 == 1]
print("\nProductos comprados: ", productos)

carnes = rules_100[rules_100['antecedents'].str.contains('CARNES CONGELADAS')]


Productos comprados:  BEBIDAS GASEOSAS     1
CARNES CONGELADAS    1
Name: 9, dtype: int64


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.



In [ ]:
fig = px.scatter(carnes, x="support", y="confidence", color="lift",
                 hover_name=carnes['antecedents'].astype(str) + ' -> ' + carnes['consequents'].astype(str),
                 labels={"support": "Support", "confidence": "Confidence", "lift": "Lift"})

fig.show()

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning:

`should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.

